Files/BaseLayout.cs

```diff
@@ -259,78 +273,141 @@ private void UnloadMenuFlyoutItemByName(string nameToUnload)
+                return (menuItem, menuHandle);
             }
 
-            return (null, null, null, null);
+            return (null, null);
         }
 
         private async void MenuLayoutItem_Click(object sender, RoutedEventArgs e)
         {
-            var selectedFileSystemItems = (App.CurrentInstance.ContentPage as BaseLayout).SelectedItems;
             var currentMenuLayoutItem = (MenuFlyoutItem)sender;
             if (currentMenuLayoutItem != null)
             {
-                var (_, _, _, command) = ParseContextMenuTag(currentMenuLayoutItem.Tag);
-                if (selectedFileSystemItems.Count > 1)
+                var (menuItem, menuHandle) = ParseContextMenuTag(currentMenuLayoutItem.Tag);
+                if (App.Connection != null)
                 {
-                    foreach (var selectedDataItem in selectedFileSystemItems)
-                    {
-                        var commandToExecute = await new ShellCommandParser().ParseShellCommand(command, selectedDataItem.ItemPath);
-                        if (!string.IsNullOrEmpty(commandToExecute.command))
-                        {
-                            await Interaction.InvokeWin32Component(commandToExecute.command, commandToExecute.arguments);
-                        }
-                    }
+                    await App.Connection.SendMessageAsync(new ValueSet() {
+                        { "Arguments", "ExecAndCloseContextMenu" },
+                        { "Handle", menuHandle },
+                        { "ItemID", menuItem.ID } });
                 }
-                else if (selectedFileSystemItems.Count == 1)
-                {
-                    var selectedDataItem = selectedFileSystemItems[0] as ListedItem;
+            }
+        }
 
-                    var commandToExecute = await new ShellCommandParser().ParseShellCommand(command, selectedDataItem.ItemPath);
-                    if (!string.IsNullOrEmpty(commandToExecute.command))
-                    {
-                        await Interaction.InvokeWin32Component(commandToExecute.command, commandToExecute.arguments);
-                    }
-                }
+        public async void RightClickItemContextMenu_Closing(object sender, object e)
+        {
+            var shellContextMenuTag = (sender as MenuFlyout).Items.Where(x => x.Tag != null)
+                .Select(x => ParseContextMenuTag(x.Tag)).FirstOrDefault(x => x.menuItem != null);
+            if (shellContextMenuTag.menuItem != null && App.Connection != null)
+            {
+                await App.Connection.SendMessageAsync(new ValueSet() {
+                    { "Arguments", "ExecAndCloseContextMenu" },
+                    { "Handle", shellContextMenuTag.menuHandle } });
             }
         }
 
         public void RightClickItemContextMenu_Opening(object sender, object e)
         {
-            SetShellContextmenu();
+            var selectedFileSystemItems = App.CurrentInstance.ContentPage.SelectedItems;
```

Comment:
`SelectedItems` already is a part of `BaseLayout`, so you don't need to use full path to it.

Reply:
Right, there is probably no need to create a temp variable at all